lib/error_tracker.ex

```diff
@@ -213,4 +213,35 @@ defmodule ErrorTracker do
 
     {error, occurrence}
   end
+
+  @doc """
+  Search errors with the given filters.
+
+  Example:
+
+      ErrorTracker.search(%{status: :unresolved, reason: "undefined"})
+
+  Valid filters are:
+    - `:reason`
+    - `:status`
+    - `:source_line`
+    - `:source_function`
+  """
+  @spec search(Ecto.Queryable.t(), %{atom() => any()}) :: Ecto.Queryable.t()
+  def search(query \\ Error, filters) when is_map(filters) do
+    Enum.reduce(filters, query, &do_search/2)
+  end
+
+  def do_search({:status, status}, query) do
+    where(query, [error], error.status == ^status)
+  end
+
+  def do_search({field, value}, query) do
+    # Postgres provides the ILIKE operator which produces a case-insensitive match between two
+    # strings. SQLite3 only supports LIKE, which is case-insensitive for ASCII characters.
+    case Repo.__adapter__() do
+      Ecto.Adapters.Postgres -> where(query, [error], ilike(field(error, ^field), ^"%#{value}%"))
+      Ecto.Adapters.SQLite3 -> where(query, [error], like(field(error, ^field), ^"%#{value}%"))
+    end
```

Comment:
@odarriba I've noticed that we are doing this in multiple places. I think that it may be a good idea to create a function along the lines of:

```elixir
Repo.with_adapter(fn ->
  :postgres -> # custom postgres specific behaviour
  :sqlite -> # custom sqlite specific behaviour
end)
```

This would have two main benefits. On the one hand we wouldn't have to manually type the adapter name every time, which has already proven dangerous. On the other hand we would be able to easily find the places that call this function in the future if we want to add another database such as MySQL.

At the moment we are doing this kind of check calling our `__adapter__()` function but in other cases we call the underlying repo's directly.

What do you think?

Reply:
Agree.